src/platforms/stm32/traceswoasync.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * This file is part of the Black Magic Debug project.
+ *
+ * Based on work that is Copyright (C) 2017 Black Sphere Technologies Ltd.
+ * Copyright (C) 2017 Dave Marples <dave@marples.net>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* This file implements capture of the TRACESWO output using ASYNC signalling.
+ *
+ * ARM DDI 0403D - ARMv7M Architecture Reference Manual
+ * ARM DDI 0337I - Cortex-M3 Technical Reference Manual
+ * ARM DDI 0314H - CoreSight Components Technical Reference Manual
+ */
+
+/* TDO/TRACESWO signal comes into the SWOUSART RX pin.
+ */
+
+#include "general.h"
+#include "cdcacm.h"
+#include "platform.h"
+
+#include <libopencmsis/core_cm3.h>
+#include <libopencm3/cm3/nvic.h>
+#include <libopencm3/stm32/timer.h>
+#include <libopencm3/stm32/f1/rcc.h>
+#include <libopencm3/stm32/usart.h>
+#include <libopencm3/stm32/dma.h>
+
+#define FULL_SWO_PACKET    (64)                             /* For speed this is set to the USB transfer size */
+#define NUM_PACKETS        (192)                            /* This is an 12K buffer */
+
+/* Default line rate....this is only used for setting the UART speed if we get a request without speed set */
+#define DEFAULTSPEED       (2250000)
+
+static volatile uint32_t w;                                 /* Which packet we are currently getting */
+static volatile uint32_t r;                                 /* Which packet we are currently waiting to transmit to USB */
+static uint8_t trace_rx_buf[NUM_PACKETS*FULL_SWO_PACKET];   /* Packets arrived from the SWO interface */
+
+static uint8_t pingpong_buf[2*FULL_SWO_PACKET];             /* Packet pingpong buffer used for receiving packets */
+
+
+
+void trace_buf_drain(usbd_device *dev, uint8_t ep)
+
+{
+    static volatile char inBufDrain;
+
+    /* If we are already in this routine then we don't need to come in again */
+    if (__atomic_test_and_set (&inBufDrain, __ATOMIC_RELAXED))
+        return;
+
+    /* Attempt to write everything we buffered */
+  if ((w!=r) && (usbd_ep_write_packet(dev, ep, &trace_rx_buf[r*FULL_SWO_PACKET], FULL_SWO_PACKET)))
```

Comment:
The USB hardware also has pingpong buffers.  This may help performance if USB is a bottleneck.  I never finished adding support for this in libopencm3, it's possible that someone else has finished this by now.